src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/EnumConverter.cs

```diff
@@ -157,21 +174,29 @@ public override void Write(Utf8JsonWriter writer, T value, JsonSerializerOptions
             // If strings are allowed, attempt to write it out as a string value
             if (_converterOptions.HasFlag(EnumConverterOptions.AllowStrings))
             {
-                string original = value.ToString();
-                if (_nameCache != null && _nameCache.TryGetValue(original, out string? transformed))
+                ulong key = Unsafe.As<T, ulong>(ref value);
+
+                if (_nameCache.TryGetValue(key, out JsonEncodedText transformed))
                 {
                     writer.WriteStringValue(transformed);
                     return;
                 }
 
+                string original = value.ToString();
                 if (IsValidIdentifier(original))
                 {
-                    transformed = _namingPolicy.ConvertName(original);
+                    JavaScriptEncoder? encoder = options.Encoder;
+
+                    // We are dealing with flags since all literal values were cached during warm-up.
+                    transformed = _namingPolicy == null
+                        ? JsonEncodedText.Encode(original, encoder)
+                        : FormatEnumValue(original, encoder);
+
                     writer.WriteStringValue(transformed);
-                    if (_nameCache != null)
-                    {
-                        _nameCache.TryAdd(original, transformed);
-                    }
+
+                    // Since the value represents a valid identifier, malicious user input is not added to the cache.
```

Comment:
This isn't a true statement. If you have a `[Flags]` enum, you have exponential growth in the number of valid inputs, which allows the attacker to flood your cache and OOM the process. For example:

```cs
[Flags]
public enum MyEnum
{
    A = 1 << 0,
    B = 1 << 1,
    C = 1 << 2,
    D = 1 << 3,
    /* ... */
    Y = 1 << 24,
    Z = 1 << 25,
}
```

An attacker can send the values `0` .. `33554431` to the server, and each one of them corresponds to a valid combination of flags. Those 33 million entries will never get cleared from the cache, resulting in unbounded memory growth and eventual OOM.

The only safe way to do this is to cache constants that are defined on the enum type itself, not arbitrary bitwise combinations of those constant values.

Reply:
BTW we should probably leave a comment to this effect in the code, lest somebody inadvertently try to introduce the same optimization later on.

Reply:
Edit: should actually be 67 million, not 33 million. I forgot how to math.

Reply:
Can we set a size limit for the dictionary and also cache _some_ of the combinations? Serializing flags maybe isn't the happy-path, but it would be great not to regress perf badly for those scenarios (since they are cached now). I think 64 is a good number and it aligns with limits set in other areas of the serializer.

Reply:
Sure, seems like that'd work. 😃

The thing to avoid is allowing untrusted payloads to cause us to allocate obnoxious amounts of memory into a permanent cache. If you have some mechanism to limit it to a more reasonable number, that's goodness.